frame/migrations/src/lib.rs

```diff
@@ -0,0 +1,469 @@
+//! 5. Must prevent re-execution of migrations.
+//! 6. Must provide transactional storage semantics for migrations.
+//! 7. Must guarantee progress.
+//!
+//! ### Design
+//!
+//! Migrations are provided to the pallet through the storage item `Migrations` of type
+//! `Get<Vec<Box<dyn SteppedMigration<…`. This was done to have the most flexibility when it comes
+//! to iterating and inspecting the migrations. It also simplifies the trait bounds since all
+//! associated types of the trait must be provided by the pallet.  
+//! The actual progress of the pallet is stored in its `Cursor` storage item. This can either be
+//! [`MigrationCursor::Active`] or `Stuck`. In the active case it points to the currently active
+//! migration and stores its inner cursor. The inner cursor can then be used by the migration to
+//! store its inner state and advance. Each time when the migration returns `Some(cursor)`, it
+//! signals the pallet that it is not done yet.   
+//! The cursor is re-set on each runtime upgrade. This ensures that it starts to execute at the
+//! first migration of the vector. The pallets cursor is only ever incremented and put into `Stuck`
+//! once it encounters an error (Goal 4). Once in the stuck state, the pallet will stay stuck until
+//! it is resolved through manual intervention.  
+//! As soon as the cursor of the pallet becomes `Some(_)`; the transaction processing will be paused
+//! by returning `true` from [`ExtrinsicSuspenderQuery::is_suspended`]. This ensures that no other
+//! transactions are processed until the pallet is done (Goal 2).  
+//! `on_initialize` the pallet will load the current migration and check whether it was already
+//! executed in the past by checking for membership of its ID in the `Historic` set. Historic
+//! migrations are ignored without causing an error. Each successfully executed migration is added
+//! to this set (Goal 5).  
+//! This proceeds until no more migrations can be loaded. This causes event `UpgradeCompleted` to be
+//! emitted (Goal 1).  
+//! The execution of each migrations happens by calling [`SteppedMigration::transactional_step`].
+//! This function wraps the inner `step` function into a transactional layer to allow rollback in
+//! the error case (Goal 6).  
+//! Weight limits must be checked by the migration itself. The pallet provides a [`WeightMeter`] for
+//! that cause. The pallet may return [`SteppedMigrationError::InsufficientWeight`] at any point.
+//! The pallet will react to this with a case decision: if that migration was exclusively execute in
+//! this block, and therefore got the maximal amount of weight possible, the pallet becomes `Stuck`.
+//! Otherwise one re-attempt is done with the same logic in the next block (Goal 3).  
+//! Progress of the pallet is guaranteed by providing once: a timeout for each migration via
+//! [`SteppedMigration::max_steps`]. The pallet **ONLY** guarantees progress if this is set to
+//! sensible limits (Goal 7).
```

Comment:
```suggestion
//! In that scenario, the one of two things will happen: if that migration was exclusively executed in
//! this block, and therefore required more than the maximum amount of weight possible, the pallet becomes `Stuck`.
//! Otherwise one re-attempt is attempted with the same logic in the next block (Goal 3).  
//! Progress through the migrations is guaranteed by providing a timeout for each migration via
//! [`SteppedMigration::max_steps`]. The pallet **ONLY** guarantees progress if this is set to
//! sensible limits (Goal 7).
```